File: strands-ts/src/interventions/actions.ts
import type {
  BeforeInvocationEvent,
  BeforeToolCallEvent,
  AfterToolCallEvent,
  BeforeModelCallEvent,
  AfterModelCallEvent,
} from '../hooks/events.js'

export type LifecycleEvent =
  | BeforeInvocationEvent
  | BeforeToolCallEvent
  | AfterToolCallEvent
  | BeforeModelCallEvent
  | AfterModelCallEvent

/**
 * Allow the operation to continue unchanged.
 *
 * @param reason - Optional metadata for debugging/logging. Not shown to the model.
 *
 * @example
 * ```typescript
 * return { type: 'proceed' }
 * ```
 */
export type Proceed = { type: 'proceed'; reason?: string }

/**
 * Block the operation. On Before* events, sets event.cancel with the reason text.
 * The reason is shown to the model as the cancellation message.
 *
 * @param reason - Why the operation was blocked. Shown to the model.
 *
 * @example
 * ```typescript
 * override beforeToolCall(event: BeforeToolCallEvent): InterventionAction {
 *   if (!this.isAuthorized(event.agent.appState.get('user_id'), event.toolUse.name)) {
 *     return { type: 'deny', reason: 'User not authorized for this tool' }
 *   }
 *   return { type: 'proceed' }
 * }
 * ```
 */
export type Deny = { type: 'deny'; reason: string }

/**
 * Provide feedback to steer behavior. On beforeToolCall/beforeInvocation, sets
 * event.cancel so the model sees the feedback and adjusts. On beforeModelCall,
 * injects feedback as a user message so the model sees it on this call.
 * On afterModelCall, the response is discarded and the model retries with the
 * feedback injected as a user message.
 *
 * @param feedback - The guidance text shown to the model.
 * @param reason - Optional metadata for debugging/logging. Not shown to the model.
 *
 * @example
 * ```typescript
 * override afterModelCall(event: AfterModelCallEvent): InterventionAction {
 *   if (this.isTooVague(event.stopData?.message)) {
 *     return { type: 'guide', feedback: 'Be more specific in your response.' }
 *   }
 *   return { type: 'proceed' }
 * }
 * ```
 */
export type Guide = { type: 'guide'; feedback: string; reason?: string }

/**
 * Pause for human approval. Calls event.interrupt() to halt agent execution
 * until the user responds. Only supported on beforeToolCall.
 *
 * @param prompt - The message shown to the human for approval. Not shown to the model.
 * @param reason - Optional metadata for debugging/logging. Not shown to the model.
 *
 * @example
 * ```typescript
 * override beforeToolCall(event: BeforeToolCallEvent): InterventionAction {
 *   if (this.requiresApproval(event.toolUse.name)) {
 *     return { type: 'interrupt', prompt: `Approve ${event.toolUse.name}?` }
 *   }
 *   return { type: 'proceed' }
 * }
 * ```
 */
export type Interrupt = { type: 'interrupt'; prompt: string; reason?: string }

/**
 * Modify event content in-place. The `apply` function mutates the event before
 * execution proceeds. Later handlers in the pipeline see the transformed content.
 *
 * The handler already has the typed event from its lifecycle method, so `apply`
 * can close over it directly — no cast needed:
 *
 * @param apply - Function that mutates the event. Not shown to the model.
 * @param reason - Optional metadata for debugging/logging. Not shown to the model.
 *
 * @example
 * ```typescript
 * override beforeToolCall(event: BeforeToolCallEvent): InterventionAction {
 *   const redacted = redactPII(event.toolUse.input)
 *   return {
 *     type: 'transform',
 *     apply: () => { event.toolUse.input = redacted },
 *     reason: 'PII redacted from tool input',
 *   }
 * }
 * ```
 */
export type Transform = { type: 'transform'; apply: (event: LifecycleEvent) => void; reason?: string }

/**
 * Union of all intervention actions a handler can return.
 *
 * | Action    | beforeInvocation | beforeToolCall | beforeModelCall | afterToolCall | afterModelCall |
 * |-----------|------------------|----------------|-----------------|---------------|----------------|
 * | Proceed   | —                | —              | —               | —             | —              |
 * | Deny      | cancel           | cancel         | cancel          | —             | —              |
 * | Guide     | cancel+          | cancel+        | inject          | —             | inject + retry |
 * | Interrupt | —                | interrupt      | —               | —             | —              |
 * | Transform | apply            | apply          | apply           | apply         | apply          |
 *
 * — = no-op (logged in audit trail, warns at runtime)
 * cancel = sets event.cancel, short-circuits (remaining handlers skipped)
 * cancel+ = sets event.cancel with accumulated feedback from all guiding handlers
 * interrupt = calls event.interrupt() for native pause/resume (human-in-the-loop)
 * inject = appends accumulated feedback as a user message so the model sees it on this call
 * inject + retry = appends accumulated feedback and retries so the model sees guidance
 * apply = calls action.apply(event) for in-place mutation, later handlers see the change
 */
export type InterventionAction = Proceed | Deny | Guide | Interrupt | Transform

/** Allow the operation to continue. */
export function proceed(reason?: string): Proceed {
  return { type: 'proceed', ...(reason !== undefined && { reason }) }
}

/** Block the operation. */
export function deny(reason: string): Deny {
  return { type: 'deny', reason }
}

/** Provide feedback to steer behavior. */
export function guide(feedback: string, reason?: string): Guide {
  return { type: 'guide', feedback, ...(reason !== undefined && { reason }) }
}

/** Pause for human approval. */
export function interrupt(prompt: string, reason?: string): Interrupt {
  return { type: 'interrupt', prompt, ...(reason !== undefined && { reason }) }
}

/** Modify event content in-place. */
export function transform(apply: (event: LifecycleEvent) => void, reason?: string): Transform {
  return { type: 'transform', apply, ...(reason !== undefined && { reason }) }
}

File: strands-py/src/strands/interventions/actions.py
"""Intervention action types.

Each action represents a typed decision that a handler returns after evaluating
an event. The framework uses these to compose decisions across multiple handlers.
"""

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from ..hooks.events import (
    AfterModelCallEvent,
    AfterToolCallEvent,
    BeforeInvocationEvent,
    BeforeModelCallEvent,
    BeforeToolCallEvent,
)

LifecycleEvent = (
    BeforeInvocationEvent | BeforeToolCallEvent | AfterToolCallEvent | BeforeModelCallEvent | AfterModelCallEvent
)

_APPROVED_RESPONSES = {"y", "yes"}


def default_evaluate(response: Any) -> bool:
    """Default evaluate function for the confirm action.

    Accepts: True, 'y'/'yes' (case-insensitive, whitespace-trimmed).

    Args:
        response: The human's response value to evaluate.

    Returns:
        True if the response is considered an approval, False otherwise.
    """
    if response is True:
        return True
    if isinstance(response, str):
        return response.lower().strip() in _APPROVED_RESPONSES
    return False


@dataclass(frozen=True)
class Proceed:
    """Allow the operation to continue unchanged.

    Args:
        reason: Optional metadata for debugging/logging. Not shown to the model.
    """

    type: str = field(default="proceed", init=False)
    reason: str | None = None


@dataclass(frozen=True)
class Deny:
    """Block the operation. The reason is shown to the model as the cancellation message."""

    type: str = field(default="deny", init=False)
    reason: str = ""


@dataclass(frozen=True)
class Guide:
    """Provide feedback to steer behavior.

    On beforeToolCall/beforeInvocation, sets cancel so the model sees the feedback.
    On beforeModelCall, injects feedback as a user message.
    On afterModelCall, the response is discarded and the model retries with feedback.

    .. warning::
        On ``after_model_call``, Guide triggers a model retry. Handlers **must** ensure
        convergence (e.g., by tracking retry count and escalating to Deny after repeated
        failures). The framework imposes no retry cap on guide-triggered retries.

    .. note::
        On ``before_model_call`` and ``after_model_call``, guidance messages are injected
        directly into ``agent.messages`` and bypass session management. Session managers
        will not track these injected messages.
    """

    type: str = field(default="guide", init=False)
    feedback: str = ""
    reason: str | None = None


@dataclass(frozen=True)
class Confirm:
    """Request human approval before proceeding. Only supported on beforeToolCall.

    Two modes depending on whether response is provided:
    - With response: passed as a preemptive value to the interrupt system, agent never pauses.
    - Without response: breaks out of the agent loop to pause for external resume.
    """

    type: str = field(default="confirm", init=False)
    prompt: str = ""
    reason: str | None = None
    response: Any = None
    evaluate: Callable[[Any], bool] | None = field(default=default_evaluate)


@dataclass(frozen=True)
class Transform:
    """Modify event content in-place.

    The apply function mutates the event before execution proceeds.
    Later handlers in the pipeline see the transformed content.
    """

    type: str = field(default="transform", init=False)
    apply: Callable[[LifecycleEvent], None] = field(default=lambda e: None)
    reason: str | None = None


InterventionAction = Proceed | Deny | Guide | Confirm | Transform
"""Union of all intervention actions a handler can return.

Action-to-event compatibility matrix::

    | Action    | before_invocation | before_tool_call | before_model_call | after_tool_call | after_model_call |
    |-----------|-------------------|------------------|-------------------|-----------------|------------------|
    | Proceed   | —                 | —                | —                 | —               | —                |
    | Deny      | cancel            | cancel           | cancel            | —               | —                |
    | Guide     | cancel+           | cancel+          | inject            | —               | inject + retry   |
    | Confirm   | —                 | confirm          | —                 | —               | —                |
    | Transform | apply             | apply            | apply             | apply           | apply            |

    — = no-op (warns at runtime)
    cancel = sets event.cancel/cancel_tool, short-circuits (remaining handlers skipped)
    cancel+ = sets cancel with accumulated feedback from all guiding handlers
    confirm = uses preemptive response or interrupt, checks with evaluate, sets cancel if denied
    inject = appends accumulated feedback as a user message so the model sees it on this call
    inject + retry = appends accumulated feedback and retries so the model sees guidance
    apply = calls action.apply(event) for in-place mutation, later handlers see the change
"""
